P5-190: Dependability (#128)
aau-hcl-p5/main: report/4.Design/Dependability.tex

\section{Dependability}\label{Design:Dependability}

> 

Reviewer: plz add a one line 'intro'

\subsection{Fault Tolerance - Bit-flipping}
As a proof of concept, the group decided to implement bit-flip fault tolerance in the laser module on the NXT, given the following rules.

\begin{enumerate}

>   \item Every variable \texttt{x} must be duplicated \texttt{x0} \cite{errorDetectionSoft}

Reviewer: ```suggestion
  \item Every variable \texttt{x} must have duplicated variable \texttt{x0} \cite{errorDetectionSoft}
```

  \item Every write operation perfomed on \texttt{x}, must also be performed on \texttt{x0} \cite{errorDetectionSoft}
  \item After each read of \texttt{x}, consistency must be checked between \texttt{x} and \texttt{x0} \cite{errorDetectionSoft}

>   \item For every conditional, repeat test in every target (both then and else branches) \cite{errorDetectionSoft}

Reviewer 2: ```suggestion
  \item For every conditional, repeat test in every conditional branch (both then and else branches) \cite{errorDetectionSoft}
```

Reviewer 2: (Full disclosure, I don't fully know, I just think this makes sense.)

\end{enumerate}

> Following the four rules above, the code could look like below, taking care of reading from variables, if-conditions and variable duplication.

Reviewer: Noget ala:
Following the four rules above, the code in Snippet~[NAVN], could be improved as shown in snippet~[NAVN2]

\noindent\begin{minipage}{.45\textwidth}

> \begin{lstlisting}[caption={Without bitflip-security},frame=tlrb,numbers=none]

Reviewer: ```suggestion
\begin{lstlisting}[caption={Without~bitflip-security},frame=tlrb,numbers=none]
```

> \begin{lstlisting}[caption={Without bitflip-security},frame=tlrb,numbers=none]

Reviewer 2: Here we have actual implementation details in the design chapter. Is this not wrong?

Author: Måske.. Det er et KORT uddrag, for at vise konceptet..

void toggle_laser() {
  if (!laser_state) {
    enable_laser();
  } else {
    disable_laser();
  }
}
\end{lstlisting}
\end{minipage}\hfill
\begin{minipage}{.45\textwidth}
\begin{lstlisting}[caption={With bitflip-security},frame=tlrb,numbers=none]
void toggle_laser() {
  if (!laser_state) {
    // Rule 4, conditional repetition
    if (laser_state) bitflip_error();
    // Rule 1+3, reading from variable.
    if (laser_state != laser_state_0) bitflip_error();

    enable_laser();
  } else {
    // Rule 4, conditional repetition
    if (!laser_state) bitflip_error();
    // Rule 1+3, reading from variable.
    if (laser_state != laser_state_0) bitflip_error();

    disable_laser();
  }
}
\end{lstlisting}
\end{minipage}

> The primary reason for choosing the laser-module, is that it in a real life sitation will be the most system critical component on the NXT.

Reviewer: ```suggestion
The primary reason for implementing bitflip protection in the laser-module, is that it will be the most system critical component on the NXT, in a real life situation.
```

\section{Dependability}\label{Design:Dependability}
In software systems, dependability has to be planned for, to a certain degree depending on a system.
This section will enlighten the thoughts behind the critical parts of the system, and try to enlighten the measures taken.

\subsection{Fault Tolerance - Bit-flipping}
As a proof of concept, it was decided to implement bit-flip fault tolerance in the laser module on the NXT, given the following rules.

\begin{enumerate}
  \item Every variable \texttt{x} must have duplicated variable \texttt{x0} \cite{errorDetectionSoft}
  \item Every write operation perfomed on \texttt{x}, must also be performed on \texttt{x0} \cite{errorDetectionSoft}
  \item After each read of \texttt{x}, consistency must be checked between \texttt{x} and \texttt{x0} \cite{errorDetectionSoft}
  \item For every conditional, repeat test in every conditional branch (both then and else branches) \cite{errorDetectionSoft}
\end{enumerate}

> Following the four rules above, the code in Snippet~\ref{snip:without_bitflip} and Snippet~\ref{snip:with_bitflip} look like below, taking care of reading from variables, if-conditions and variable duplication.

Reviewer 3: Det er basically at for hver variabel x har du en ekstra variabel x0 med samme værdi, det sikrer at hvis den ene bliver bitflipped tjekker man den anden om den har samme værdi

Hvis de ikke har det er vi ret sikre på at der et sket et bitflip eller lignende
